src/napari_matplotlib/base.py

```diff
@@ -69,25 +73,55 @@ def add_single_axes(self) -> None:
         self.axes = self.figure.subplots()
         self.apply_napari_colorscheme(self.axes)
 
-    @staticmethod
-    def apply_napari_colorscheme(ax: Axes) -> None:
+    def apply_napari_colorscheme(self, ax: Axes) -> None:
         """Apply napari-compatible colorscheme to an Axes."""
+        # get the foreground colours from current theme
+        theme = napari.utils.theme.get_theme(self.viewer.theme, as_dict=False)
+        fg = theme.foreground.as_hex()  # fg is a muted contrast to bg
+        tx = theme.text.as_hex()  # text is high contrast to bg
+
         # changing color of axes background to transparent
         ax.set_facecolor("none")
 
         # changing colors of all axes
         for spine in ax.spines:
-            ax.spines[spine].set_color("white")
+            ax.spines[spine].set_color(fg)
 
-        ax.xaxis.label.set_color("white")
-        ax.yaxis.label.set_color("white")
+        ax.xaxis.label.set_color(tx)
+        ax.yaxis.label.set_color(tx)
 
         # changing colors of axes labels
-        ax.tick_params(axis="x", colors="white")
-        ax.tick_params(axis="y", colors="white")
+        ax.tick_params(axis="x", colors=tx)
+        ax.tick_params(axis="y", colors=tx)
+
+    def _on_theme_change(self) -> None:
+        """Update MPL toolbar and axis styling when `napari.Viewer.theme` is changed.
+
+        Note:
+            At the moment we only handle the default 'light' and 'dark' napari themes.
+        """
+        self._replace_toolbar_icons()
+        if self.figure.gca():
+            self.apply_napari_colorscheme(self.figure.gca())
```

Comment:
Not sure if this is quite the correct line, but I still don't quite have a perfect UX when switching the theme whilst the widgets are open...

https://github.com/matplotlib/napari-matplotlib/assets/1836192/5453c2e7-e857-4424-9a76-6adb33d4f757

Reply:
A bit of a dive into the `napari` source, and you might want to use something like this line to trigger something when the theme is changed? 

https://github.com/napari/napari/blob/46dcc9c0976c35ec9dc64be60235f4addcadaa5c/napari/_qt/qt_main_window.py#L578